telegram/files/file.py

```diff
@@ -90,15 +89,15 @@ def __init__(
         self.file_unique_id = str(file_unique_id)
         # Optionals
         self.file_size = file_size
-        self.file_path = file_path
+        self.file_path: Optional[str] = file_path  # Can be either path or url, so must remain str.
```

Comment:
I don't quite understand this change. Also, as `File` is defined by the TG Bot API, I'd argue that we should not allow `file_path` to be something else than `None` or a string

Reply:
isn't this what the type annotation denotes? that this variables can either be ```str``` or None?

Reply:
ah, I thought with "can be path" you meant `pathlib.Path`. Then this is the same as https://github.com/python-telegram-bot/python-telegram-bot/pull/2688#discussion_r720698609 and https://github.com/python-telegram-bot/python-telegram-bot/pull/2688#discussion_r720698795.

you should only need to annotate something if a type checker (here: mypy, because this is the type checker that we use in the pre-commit hooks) can't reasonably infer the type on its own. In this example, `file_path` is already annotated as `Optional[str]`, so by `self.file_path = file_path` it's clear that `self.file_path` has the same type.
Adding annotations explicitly where not needed may improve readability, but I guess that's a bit matter of taste. Personally I'd advocate to use as few annotations as possible because every annotation that you make means that the type checker doesn't understand your code, meaning the code is complex meaning that it potentionally has a bug :)

https://github.com/python-telegram-bot/python-telegram-bot/pull/2688#discussion_r720698609 is in fact a bit different, because here the return value depends on the input. We could actually improve this with the `typing.overload` decorator. If you're unfamiliar with that feel free to ping me, then I'll have a look at it :)

Reply:
ok no problem, thank you for the elaborated explanation on why you advocate to refrain from using type annotations when not needed. Which leads me to the question - how do i know when type annotations are needed and that i should add them?

Personally, i prefer to include type annotations to all instance attributes when defining them in an ```__init__``` method, because i find it helps readability when reading code, for quick understanding of the class's abilities.
However, i did not take under consideration the perspective that sees an explicit type annotation as a last resurt after the type checker failed to understand what is the type at hand.
So, i removed type annotation here. :relaxed:

Reply:
upon pushing changes, i see that test-pre-commit fails now.
Could it be related to me removing the type annotation?
```
examples/arbitrarycallbackdatabot.py:87: error: No overload variant of "PicklePersistence" matches argument type "str"  [call-overload]
examples/arbitrarycallbackdatabot.py:87: note: Possible overload variants:
examples/arbitrarycallbackdatabot.py:87: note:     def [UD, CD, BD] __init__(self, filepath: Union[Path, str], store_data: PersistenceInput = ..., single_file: bool = ..., on_flush: bool = ...) -> PicklePersistence[Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]]
examples/arbitrarycallbackdatabot.py:87: note:     def [UD, CD, BD] __init__(self, filepath: Union[Path, str], store_data: PersistenceInput = ..., single_file: bool = ..., on_flush: bool = ..., context_types: ContextTypes[Any, UD, CD, BD] = ...) -> PicklePersistence[UD, CD, BD]
examples/persistentconversationbot.py:135: error: No overload variant of "PicklePersistence" matches argument type "str"  [call-overload]
examples/persistentconversationbot.py:135: note: Possible overload variants:
examples/persistentconversationbot.py:135: note:     def [UD, CD, BD] __init__(self, filepath: Union[Path, str], store_data: PersistenceInput = ..., single_file: bool = ..., on_flush: bool = ...) -> PicklePersistence[Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]]
examples/persistentconversationbot.py:135: note:     def [UD, CD, BD] __init__(self, filepath: Union[Path, str], store_data: PersistenceInput = ..., single_file: bool = ..., on_flush: bool = ..., context_types: ContextTypes[Any, UD, CD, BD] = ...) -> PicklePersistence[UD, CD, BD]
Found 2 errors in 2 files (checked 18 source files)
```

Reply:
>  Which leads me to the question - how do i know when type annotations are needed and that i should add them?

I'd say ideally you should only need to add annotations in a function signature. Everywhere else you only need them, if mypy complains. the only case where I recall needing manual annotations is something like

```python
if condition:
    var: Optional[str] = 'foo'
else:
    var = None
```
Here the annotation is needed because otherwise would complain about `var = None` with "you're assigning `None` to a variable of type `str`"